lib/block_view_lockups.go

```diff
@@ -0,0 +1,1784 @@
+	tombstoneLockedBalanceEntry := *lockedBalanceEntry
+	tombstoneLockedBalanceEntry.isDeleted = true
+
+	// Set the LockupYieldCurvePoint as deleted in the view.
+	bav._setLockedBalanceEntry(&tombstoneLockedBalanceEntry)
+}
+
+// Get Helper Functions for LockedBalanceEntry
+
+func (bav *UtxoView) GetLockedBalanceEntryForHODLerPKIDProfilePKIDExpirationTimestampNanoSecs(
+	hodlerPKID *PKID, profilePKID *PKID, expirationTimestampNanoSecs int64) (_lockedBalanceEntry *LockedBalanceEntry) {
+	// Create a key associated with the LockedBalanceEntry.
+	lockedBalanceEntryKey := (&LockedBalanceEntry{
+		HODLerPKID:                  hodlerPKID,
+		ProfilePKID:                 profilePKID,
+		ExpirationTimestampNanoSecs: expirationTimestampNanoSecs,
+	}).ToMapKey()
+
+	// Check if the key exists in the view.
+	if viewEntry, viewEntryExists :=
+		bav.LockedBalanceEntryKeyToLockedBalanceEntry[lockedBalanceEntryKey]; viewEntryExists {
+		return viewEntry
+	}
+
+	// No mapping exists in the view, check for an entry in the DB.
+	lockedBalanceEntry := DBGetLockedBalanceEntryForHODLerPKIDProfilePKIDExpirationTimestampNanoSecs(
+		bav.Handle, bav.Snapshot, hodlerPKID, profilePKID, expirationTimestampNanoSecs)
+
+	// Cache the DB entry in the in-memory map.
+	if lockedBalanceEntry != nil {
+		bav._setLockedBalanceEntry(lockedBalanceEntry)
+	}
+
+	return lockedBalanceEntry
+}
+
+func (bav *UtxoView) GetUnlockableLockedBalanceEntries(
+	hodlerPKID *PKID,
+	profilePKID *PKID,
+	currentTimestampNanoSecs int64,
+) ([]*LockedBalanceEntry, error) {
+	// Validate inputs.
+	if hodlerPKID == nil {
+		return nil, errors.New("UtxoView.GetUnlockableLockedBalanceEntries: nil hodlerPKID provided as input")
+	}
+	if profilePKID == nil {
+		return nil, errors.New("UtxoView.GetUnlockableLockedBalanceEntries: nil profilePKID provided as input")
+	}
+
+	// First, pull unlockable LockedBalanceEntries from the db and cache them in the UtxoView.
+	dbUnlockableLockedBalanceEntries, err := DBGetUnlockableLockedBalanceEntries(
+		bav.Handle, bav.Snapshot, hodlerPKID, profilePKID, currentTimestampNanoSecs)
+	if err != nil {
+		return nil, errors.Wrapf(err, "UtxoView.GetUnlockableLockedBalanceEntries")
+	}
+	for _, lockedBalanceEntry := range dbUnlockableLockedBalanceEntries {
+		// Cache results in the UtxoView.
+		if _, exists := bav.LockedBalanceEntryKeyToLockedBalanceEntry[lockedBalanceEntry.ToMapKey()]; !exists {
+			bav._setLockedBalanceEntry(lockedBalanceEntry)
+		}
+	}
+
+	// Then, pull unlockable LockedBalanceEntries from the UtxoView.
+	var unlockableLockedBalanceEntries []*LockedBalanceEntry
+	for _, lockedBalanceEntry := range bav.LockedBalanceEntryKeyToLockedBalanceEntry {
+		// Filter to matching LockedBalanceEntries.
+		if !lockedBalanceEntry.HODLerPKID.Eq(hodlerPKID) ||
+			!lockedBalanceEntry.ProfilePKID.Eq(profilePKID) ||
+			lockedBalanceEntry.ExpirationTimestampNanoSecs > currentTimestampNanoSecs ||
+			lockedBalanceEntry.BalanceBaseUnits.IsZero() ||
```

Comment:
Hmm... isn't it possible that the balance could be zero? If this is possible, then we would end up never garbage-collecting a zero entry, resulting in bugs. Consider removing the IsZero() check here.

Reply:
Currently we garbage collect on the flush. Not sure if this is what we want to do, but it's something we did with BalanceEntry as well. It's pretty deep in the DB utils where we actually garbage collect the zero balances.